discovery/aws/ec2_test.go

```diff
@@ -0,0 +1,467 @@
+	labels := model.LabelSet{
+		"__address__":                     model.LabelValue(fmt.Sprintf("%s:%d", *ec2Data.instances[0].PrivateIpAddress, port)),
+		"__meta_ec2_ami":                  model.LabelValue(*ec2Data.instances[0].ImageId),
+		"__meta_ec2_architecture":         model.LabelValue(*ec2Data.instances[0].Architecture),
+		"__meta_ec2_availability_zone":    model.LabelValue(ec2Data.azNames[azIdx]),
+		"__meta_ec2_availability_zone_id": model.LabelValue(ec2Data.azIDs[azIdx]),
+		"__meta_ec2_instance_id":          model.LabelValue(*ec2Data.instances[0].InstanceId),
+		"__meta_ec2_instance_lifecycle":   model.LabelValue(*ec2Data.instances[0].InstanceLifecycle),
+		"__meta_ec2_instance_state":       model.LabelValue(*ec2Data.instances[0].State.Name),
+		"__meta_ec2_instance_type":        model.LabelValue(*ec2Data.instances[0].InstanceType),
+		"__meta_ec2_owner_id":             model.LabelValue(ec2Data.ownerID),
+		"__meta_ec2_platform":             model.LabelValue(*ec2Data.instances[0].Platform),
+		"__meta_ec2_private_dns_name":     model.LabelValue(*ec2Data.instances[0].PrivateDnsName),
+		"__meta_ec2_private_ip":           model.LabelValue(*ec2Data.instances[0].PrivateIpAddress),
+		"__meta_ec2_public_dns_name":      model.LabelValue(*ec2Data.instances[0].PublicDnsName),
+		"__meta_ec2_public_ip":            model.LabelValue(*ec2Data.instances[0].PublicIpAddress),
+		"__meta_ec2_region":               model.LabelValue(ec2Data.region),
+	}
+
+	for i := 0; i < len(ec2Data.instances[0].Tags); i++ {
+		key := strings.ReplaceAll(fmt.Sprintf("__meta_ec2_tag_%s", *ec2Data.instances[0].Tags[i].Key), "-", "_")
+		labels[model.LabelName(key)] = model.LabelValue(*ec2Data.instances[0].Tags[i].Value)
+	}
+
+	return labels
+}
+
+// The tests itself.
+func TestMain(m *testing.M) {
+	goleak.VerifyTestMain(m)
+}
+
+func TestRefreshAZIDs(t *testing.T) {
+	ctx := context.Background()
+	client := &mockEC2Client{}
+
+	GenerateTestAzData()
+
+	d := &EC2Discovery{
+		ec2: client,
+	}
+
+	err := d.refreshAZIDs(ctx)
+	require.NoError(t, err)
+	require.Equal(t, ec2Data.azToAZID, d.azToAZID)
+}
+
+func TestRefreshAZIDsHandleError(t *testing.T) {
+	ctx := context.Background()
+	client := &mockEC2Client{}
+
+	ec2Data = EC2Data{}
+
+	d := &EC2Discovery{
+		ec2: client,
+	}
+
+	err := d.refreshAZIDs(ctx)
+	require.Equal(t, "No AZs found", err.Error())
+}
+
+func TestRefreshNoPrivateIp(t *testing.T) {
+	ctx := context.Background()
+	client := &mockEC2Client{}
+
+	GenerateTestAzData()
+
+	instance := ec2.Instance{}
+	instance.SetInstanceId(fmt.Sprintf("instance-id-%s", RandString(8)))
+
+	ec2Data.instances = append(ec2Data.instances, &instance)
+
+	d := &EC2Discovery{
+		ec2: client,
+		cfg: &EC2SDConfig{
+			Region: ec2Data.region,
+		},
+	}
+
+	expected := make([]*targetgroup.Group, 1)
+	expected[0] = &targetgroup.Group{
+		Source: ec2Data.region,
+	}
+
+	g, err := d.refresh(ctx)
+	require.Equal(t, expected, g)
+	require.NoError(t, err)
```

Comment:
nit: please notice that all your tests end up asserting these same conditions.  

I think you can refactor the tests for the `refresh` function specifically to be table driven. https://go.dev/wiki/TableDrivenTests

The idea is that you build a table where each row contains a test name, the input data for the tests and optionally some data used in assertions. This way it is easier to add new test cases.

Reply:
I did something similar in https://github.com/prometheus/prometheus/pull/14364/commits/5b66987fb14c0a20507fb03a69af36ae79ec5e39 but I still had to use generator functions to create the input and the expected output. But the list of tests is more manageable this way. I think.